src/docs/spec-formats-tsv.md

```diff
@@ -0,0 +1,274 @@
+
+Implementations SHOULD support reading SSSOM/TSV files in external metadata mode; they MAY support writing SSSOM/TSV files in that mode.
+
+
+## Encoding
+
+SSSOM/TSV files MUST be encoded in UTF-8 ([RFC 3629](https://datatracker.ietf.org/doc/html/rfc3629#section-13)). They MUST NOT start with a byte order mark (U+FEFF). This applies to external metadata files as well, when the [external metadata mode](#external-metadata-mode) is used.
+
+
+## Identifiers
+
+All identifiers in a SSSOM/TSV file, that is, all the values of slots typed as [EntityReference](EntityReference.md), MUST be serialised in [CURIE syntax](https://www.w3.org/TR/curie/). SSSOM/TSV parsers SHOULD reject files containing identifiers serialised as IRIs.
+
+To allow unambiguous resolution of all CURIEs present in a SSSOM/TSV file, the metadata block MUST contain an additional `curie_map` field, which is a map of prefix names to IRI prefixes. The `curie_map` field SHOULD appear at the beginning of the metadata block.
+
+Any prefix name used in a SSSOM/TSV file MUST be declared with a corresponding entry in the CURIE map. SSSOM/TSV parsers MUST reject a file with undeclared prefix names.
+
+Prefix names listed in the table found in the [IRI prefixes](spec-intro.md#iri-prefixes) section are considered “built-in”. As such, they MAY be omitted from the CURIE map. If they are not omitted, they MUST point to the same IRI prefixes as in the aforementioned table.
+
+
+## Propagatable slots
+
+As [explained in another section](spec-model.md#propagation-of-mapping-set-slots), some slots in the `MappingSet` class are intended, not to describe the mapping set itself, but to store values that are shared by all mappings in the set. These slots are called the “propagatable slots”, because their values should be “propagated” from the mapping set down to the individual mappings.
+
+### Propagation
+
+“Propagation” is the operation of assigning to individual mappings in a set the values from the propagatable slots of the set. That operation SHOULD be performed by a SSSOM/TSV parser before passing the parsed objects to the application code.
+
+For any given propagatable slot, propagation is only allowed if none of the individual mappings already have their own value in that slot. If any mapping (even only one mapping) has a value in that slot, then the slot MUST be considered as non-propagatable. Otherwise, a propagating SSSOM/TSV parser MUST (1) copy over the value of the propagatable slot on the `MappingSet` object to the corresponding slot of every individual `Mapping` objects, and (2) remove the propagated value from the `MappingSet` object.
+
+### Condensation
+
+“Condensation” is the opposite of “propagation”. It is the operation of assigning common values to the propagatable slots of the set, based on the values of these slots on individual mappings. That operation SHOULD be performed a SSSOM/TSV writer prior to writing a set into a SSSOM/TSV file.
```

Comment:
It may be “new” in that we did not discuss it in #305, but it does not come from nowhere. This is the logical consequence of allowing propagation, since it is the opposite operation. You would have nothing to propagate if there wasn’t any condensation to begin with.

> In most cases, people wont want to condensate

Why? That’s an internal detail of the format. Why would people care about how the metadata are stored, as long as the metadata _are_ stored and can be retrieved whenever needed?

That would be like saying that people using ZIP files won’t want their data to be compressed with the DEFLATE algorithm. Why would they care?

Reply:
I dont know if I agree with the analogy here. I am not so concerned about SHOULD vs MAY, but I am concerned that tools will take the SHOULD as a permission to build serialisers that do not even _permit_ un-normalised / un-condensed outputs.. I want to be clear that if a user _wants_ to serialise an uncondensed version, they should be able to.

Reply:
But even so: what would be the problem if you have a tool that only generate condensed output?

First, for those who craft their sets “manually”, they would still be able to write them in whatever way they want. Readers that support reading condensed files (that is, readers that supports propagation, which should be most readers since propagation is a SHOULD) will always support reading uncondensed files.

And for those who create sets using a tool that only generate condensed output: again, what is the problem? As long as readers SHOULD support propagation (which they really should, otherwise as I said further below, the whole idea of having propagatable slots becomes useless), condensed values will always be propagated back upon reading. Users won’t even need to know that the file has been written in condensed form.

Reply:
> I am concerned that tools will take the SHOULD as a permission to build serialisers that do not even permit un-normalised / un-condensed outputs..

I could understand if the concern was that implementers take the SHOULD as a permission to build _readers_ that do not support reading anything other than canonical and condensed output (“since writers SHOULD write condensed output, we should encounter very few or even zero files that are non-condensed, so we don’t need to bother supporting them”). That would be bad indeed.

But there is no risk of that.

First, it is impossible to support reading condensed files _only_. A reader that can read condensed files (that is: a reader that can perform propagation, to “un-condense” the slots that were condensed) can necessarily also read non-condensed files.

Second, regarding canonicalisation: the proposed spec explicitly forbids (MUST NOT) a reader from rejecting a file solely because it is not in the canonical format.

Reply:
How about:

> Writers SHOULD perform condensation before writing a set, but that behaviour MUST be deactivatable.

?

Reply:
If you are fine with this, then great, that covers my concerns fully!

Reply:
Deactivating condensation in SSSOM-Java's TSVWriter is a matter of putting _one_ function call behind a conditional expression, so if it's all it takes to get you to agree that condensation SHOULD be supported, I'm a fine with that. :)

I'll also add something like

> Condensation and propagation work in concert. Therefore, an implementation that supports one operation MUST support the other.

This doesn't change the fact that both operations remain a "SHOULD" requirement.